src/ripple/protocol/SeqOrTicket.h

```diff
@@ -0,0 +1,157 @@
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef RIPPLE_PROTOCOL_SEQ_OR_TICKET_H_INCLUDED
+#define RIPPLE_PROTOCOL_SEQ_OR_TICKET_H_INCLUDED
+
+#include <cstdint>
+#include <ostream>
+
+namespace ripple {
+
+/** A type that represents either a sequence value or a ticket value.
+
+  We use the value() of a SeqOrTicket in places where a sequence was used
+  before.  An example of this is the sequence of an Offer stored in the
+  ledger.  We do the same thing with the in-ledger identifier of a
+  Check, Payment Channel, and Escrow.
+
+  Why is this safe?  If we use the SeqOrTicket::value(), how do we know that
+  each ledger entry will be unique?
+
+  There are two components that make this safe:
+
+  1. When an account creates tickets it must use a sequence number.  You may
+     not use a ticket to create tickets.
+
+     Because the account used that sequence to create the ticket, we know that
+     for the given account the sequence was consumed by the "TicketCreate"
+     transaction.  Since a sequence can only be used once, we know that
+     sequence cannot be used for an offer or any other transaction.  It was
+     used by the "TicketCreate"
+
+     Since the sequence number on the ticket was previously used to build a
+     ticket, that same sequence number cannot have previously been used to
+     create (for example) an Offer on the same account.  So as long as we make
+     sure a ticket can only be consumed by one transaction, we can be
+     confident that using the ticket number to fill in the sequence of an
+     Offer would not duplicate a sequence used on another Offer created by
+     the same account.
+
+  2. When a "TicketCreate" transaction creates a batch of tickets it advances
+     the account root sequence to one past the largest created ticket.
+
+     Therefore all tickets in a batch other than the first may never have
+     the same value as a sequence on that same account.  And since a ticket
+     may only be used once there will never be any duplicates within this
+     account.
+*/
+class SeqOrTicket
+{
+public:
+    enum Type : std::uint8_t
+    {
+        seq = 0,
+        ticket
+    };
+
+private:
+    std::uint32_t value_;
+    Type type_;
+
+public:
+    constexpr explicit SeqOrTicket (Type t, std::uint32_t v)
```

Comment:
_Almost_ every situation that wants a `Sequence` can also take a `Ticket`.  There are a few exceptions.  But I started out by trying to handle them as separate concepts and it was not working.

Every valid transaction comes in with exactly one of `Sequence` or `Ticket`. You could put three interfaces on a transaction that return `SeqOrTicket`, `Sequence`, and `Ticket`.  But there's no _static_ safety there.  The best you can do is throw an exception if the wrong type is asked for.  And _usually_ you don't have to tell them apart.  So providing a single interface that can always be satisfied (a `SeqOrTicket`) seemed like the right answer.  And I think it plays out well in the code.

I could try having a conversion operator hierarchy like we have with `TEC` codes.  Then we'd have `Sequence`, `Ticket`, and `SeqOrTicket`.  But we would have very few uses of `Sequence` or `Ticket`.  And for the most part you'd get to `Sequence` or `Ticket` with a runtime check followed by an explicit cast.  I suspect it would not improve the situation much.  But I could be wrong.

Reply:
I would not have a conversion hierarchy - I agree, that's overkill. I'll take a closer look at making sequence and ticket distinct types, you might be right that it's not worth it.